### PR TITLE
Fix bug#227

### PR DESCRIPTION
**Pull request by Sam1301**
This pr addresses #227 
The server url first entered (when the app is started) was being used every time, regardless of what the user typed in next. Therefore, I added a new function `setZulipServices()` in `ZulipApp.java` and used it to set `zulipServices` to `null` before calling `getServices()` in `LoginActivity`. This will cause new `zulipServices` object creation **only during login**.
I have also hidden `/api` from the user and removed any leading/trailing spaces in `serverUrl` entered by the user.

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@Sam1301, it looks like you've already signed the Dropbox CLA. Thanks!

**Comment by timabbott** [comment]
@kunall17 @niftynei if one of you can review this, I can take care of uploading a new beta based on it for @jmandel to test.

**Comment by Sam1301** [comment]
Updated the pr to take care of some more end cases. The app crashes if the server url doesn't end with `/`. Also, the `api/` should be added at the end of server url if the paths segment is empty or the last path provided by the user is not `api/`.

File: app/src/main/java/com/zulip/android/activities/LoginActivity.java [hunk 1]
```diff
@@ -184,20 +188,31 @@ public void onSaveInstanceState(Bundle savedInstanceState){
...
         Uri serverUri = Uri.parse(serverURL);
 
         serverUri = serverUri.buildUpon().scheme(httpScheme).build();
 
-        // if does not begin with "api.zulip.com" and if the path is empty, use "/api" as first segment in the path
-        List<String> paths = serverUri.getPathSegments();
-        if (!serverUri.getHost().startsWith("api.") && paths.isEmpty()) {
-            serverUri = serverUri.buildUpon().appendEncodedPath("api/").build();
-        }
+        // display server url with http scheme used
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/LoginActivity.java]: Can you split this another commit as it has nothing to do with the issue! 
Btw why do you want to hide the /api  being appended in the server URL? 
I think it was ok to display the complete url!

File: app/src/main/java/com/zulip/android/activities/LoginActivity.java [hunk 2]
```diff
@@ -184,20 +188,31 @@ public void onSaveInstanceState(Bundle savedInstanceState){
...
+        // if server url does not end with "api/" or if the path is empty, use "/api" as last segment in the path
+        List<String> paths = serverUri.getPathSegments();
+        if (paths.isEmpty() || !paths.get(paths.size() - 1).equals("api")) {
+            serverUri = serverUri.buildUpon().appendEncodedPath("api/").build();
+        }
+
         ((ZulipApp) getApplication()).setServerURL(serverUri.toString());
 
+        // create new zulipServices object every time by setting it to null
+        getApp().setZulipServices(null);
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/LoginActivity.java]: I think we can move this line to onError and onFailure methods, as if the connection is successful why to recreate ZulipServices object again?

File: app/src/main/java/com/zulip/android/activities/LoginActivity.java (hunk 1 above)
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/LoginActivity.java]: Sure :+1:  
As mentioned by @timabbott here https://github.com/zulip/zulip-android/issues/227, we should hide 'api/' from user as old api.zulip.com doesn't exist anymore.

File: app/src/main/java/com/zulip/android/activities/LoginActivity.java (hunk 2 above)
Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/LoginActivity.java]: Yeah that makes sense, I didn't gave a thought about logout.
If you would set the ZulipServices null after logging out it would work, as the retrofit baseURL is what making the difference.
I am just trying to avoid the re initialization of the ZulipServices after logging in.

File: app/src/main/java/com/zulip/android/activities/LoginActivity.java (hunk 1 above)
Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/LoginActivity.java]: :+1:

File: app/src/main/java/com/zulip/android/activities/LoginActivity.java (hunk 2 above)
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/LoginActivity.java]: @kunall17 Can you help me with this, I cannot think about the case when `showBackends()` will be called after logging in which would lead to the unnecessary re initialization of `zulipServices`.

Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/LoginActivity.java]: Even after successful login the ZulipServices will be null, now when AsyncGetEvents calls getZulipServices  here https://github.com/zulip/zulip-android/blob/c9855cf2b4f531b325e65fe324a362be75396553/app/src/main/java/com/zulip/android/networking/AsyncGetEvents.java#L107

Those objects will be re initialized!

Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/LoginActivity.java]: Infact getServices is being called in the next line, didn't noticed this one! 
https://github.com/Sam1301/zulip-android/blob/d8eb3d7bd001482e6873109ac92ed85b0dc15817/app/src/main/java/com/zulip/android/activities/LoginActivity.java#L216

Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/LoginActivity.java]: Thanks @kunall17 for clearing this out.
I am not sure if I got your point. I just tried to debug the app and couldn't find `zulipServices` being recreated. After `zulipServices` object is initialized(during a successful login), the if condition `if (zulipServices == null)` in getZulipServices() prevents it from being initialized again.

Comment (kunall17) [inline app/src/main/java/com/zulip/android/activities/LoginActivity.java]: Ok, ya the code is fine, got confused somewhere.. :+1:

**Comment by kunall17** [comment]
@timabbott lgtm .. :+1:

File: app/src/main/java/com/zulip/android/activities/LoginActivity.java (hunk 2 above)
Comment (Sam1301) [inline app/src/main/java/com/zulip/android/activities/LoginActivity.java]: Thanks for reviewing this @kunall17 :+1:

**Comment by timabbott** [comment]
@Sam1301 can you clean up the commit messages to match the standard Zulip style?  

Check out the Zulip commit message guidelines for more details: https://zulip.readthedocs.io/en/latest/version-control.html#commit-messages

**Comment by Sam1301** [comment]
Thanks for the guide, updated the PR :+1:

**Comment by timabbott** [comment]
@Sam1301 I just went ahead and fixed this for these commits while merging this, but you need a blank line between the first line of your commit messages and the later lines.  

Merged, thanks @Sam1301!

**Comment by Sam1301** [comment]
Thanks! I'll take care of this from next time.
